### PR TITLE
Store deleted block in temp table - Closes #3772

### LINKED ISSUE #3772: Blocks need to be stored in temporary table on block deletion
### Expected behavior
Whenever we have found and verified the `commonBlock`, we will store the blocks from this commonHeight til the height of the tip in a temporary table so we can start applying the newly received blocks.

**Implement the functionality to**:
- delete blocks from the current `blocks` table
- create a backup of the deleted blocks in a temp table

### Which version(s) does this affect? (Environment, OS, etc...)
`feature/improve-bft`

### PR DESCRIPTION
**Pull request by SargeKhan**
### What was the problem?
Blocks which are deleted as part of fork recovery mechanism in BFT need to be stored in the temp table. This temp table was non existent in the application. 
### How did I solve it?
Created the required temp table, its migration and exposed this functionality from chain module.
### How to manually test it?
Run the unit tests for entity as well as the deleteLastBlockAndStoreInTemp function
### Review checklist

- [ ] The PR resolves #3772 
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

### CODE REVIEW

**Comment by SargeKhan** [comment]
I have written tests in jest while importing some libraries from mocha. This is an antipattern, i concede but i choose to favor having tests than not. Feel free to suggest a better alternative.

**Comment by SargeKhan** [comment]
Another note, the diff on github is not correct as it shows too many changes. I am listing my commits which need to be reviewed for this PR to be merged:
- declare database entity for block_temp table 322aabe
- create sql querires for crd operation on block_temp table 811e72c
- create migration for block_temp table 273e57c
- integrate block_temp entity in application startup and test faa3385
- add test for block_temp entity 384ed4a
- fix tests for deleteLastBlockAndStoreInTemp 781c314

File: framework/src/modules/chain/blocks/chain.js [hunk 1]
```diff
@@ -141,16 +141,48 @@ class BlocksChain {
...
+					this.storage,
+					this.interfaceAdapters,
+					this.genesisBlock,
+					this.roundsModule,
+					this.slots,
+					lastBlock,
+					this.exceptions,
+					tx
+				);
+				const parsedDeletedBlock = parseBlockToJson(lastBlock);
```
Comment (nazarhussain) [inline framework/src/modules/chain/blocks/chain.js]: You returned `previousBlock` but to store used `lastBlock`, any particular reason for it?

File: framework/src/modules/chain/blocks/chain.js [hunk 2]
```diff
@@ -141,16 +141,48 @@ class BlocksChain {
...
+				const blockTempEntry = {
+					id: parsedDeletedBlock.id,
+					height: parsedDeletedBlock.height,
+					fullBlock: parsedDeletedBlock,
+				};
+				await this.storage.entities.BlockTemp.create(blockTempEntry, {}, tx);
```
Comment (nazarhussain) [inline framework/src/modules/chain/blocks/chain.js]: Would be nice to have `storage.entities.TempBlock` rather `storage.entities.BlockTemp`

File: framework/src/modules/chain/blocks/chain.js [hunk 3]
```diff
@@ -195,11 +224,29 @@ const saveBlock = async (storage, block, tx) => {
...
 
-	parsedBlock.transactions.map(transaction => {
+	parsedBlock.transactions.forEach(transaction => {
 		transaction.blockId = parsedBlock.id;
 		return transaction;
 	});
 
+	parsedBlock.transactions = parsedBlock.transactions.map(transaction =>
+		transaction.toJSON()
+	);
```
Comment (nazarhussain) [inline framework/src/modules/chain/blocks/chain.js]: We should override the block id after this block 

```
	parsedBlock.transactions.forEach(transaction => {
 		transaction.blockId = parsedBlock.id;		 		
 		return transaction;
 	});

```

File: framework/src/modules/chain/components/storage/entities/block_temp.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ * @property {number} [height_eql]
+ * @property {number} [height_ne]
+ * @property {number} [height_gt]
+ * @property {number} [height_gte]
+ * @property {number} [height_lt]
+ * @property {number} [height_lte]
+ * @property {number} [height_in]
+ */
+
+class BlockTemp extends BaseEntity {
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/block_temp.js]: Please rename it to `TempBlock`

File: framework/src/modules/chain/components/storage/entities/block_temp.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	delete: 'block_temp/delete.sql',
+	get: 'block_temp/get.sql',
+};
+
+/**
+ * Block temp
+ * @typedef {Object} block_temp
+ * @property {string} id
+ * @property {number} height
+ * @property {json} block
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/block_temp.js]:
```suggestion
 * @property {json} fullBlock
```

File: framework/src/modules/chain/components/storage/entities/block_temp.js [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	 * @param {Object} data
+	 * @param {Object} [_options]
+	 * @param {Object} [tx] - Transaction object
+	 * @return {null}
+	 */
+	// eslint-disable-next-line no-unused-vars
+	create({ height, id, fullBlock }, _options = {}, tx = null) {
+		assert(height && Number.isInteger(height), 'height must be a number');
+		assert(id && typeof id === 'string', 'id must be a string');
+		assert(typeof fullBlock === 'object', 'block must be an object');
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/block_temp.js]: Typeof works good only for primitives, I would prefer to use 
``
fullBlock instanceof Object
``

File: framework/src/modules/chain/components/storage/sql/block_temp/create.sql [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+INSERT INTO block_temp (
+	${fields:raw}
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/sql/block_temp/create.sql]: As we have always fixed set of fields, then would prefer to hardcode the field names in SQL. When you have dynamic fields based on conditions then we calculate it JS and pass it as params.

File: framework/src/modules/chain/init_steps/bootstrap_storage.js [hunk 8]
```diff
@@ -37,6 +38,7 @@ module.exports = async ({ components: { storage, logger } }, accountLimit) => {
 			replaceExisting: true,
 		});
 		storage.registerEntity('ChainMeta', ChainMeta);
+		storage.registerEntity('BlockTemp', BlockTemp);
```
Comment (nazarhussain) [inline framework/src/modules/chain/init_steps/bootstrap_storage.js]: As mentioned above 
```suggestion
		storage.registerEntity('TempBlock', TempBlock);
```

File: framework/src/modules/chain/loader.js [hunk 9]
```diff
@@ -283,19 +283,12 @@ class Loader {
...
-		} catch (p2pError) {
-			this.logger.error('Failed to load block from network', p2pError);
-			return [];
-		}
+		const { data } = await this.channel.invoke('network:request', {
+			procedure: 'blocks',
+			data: {
+				lastBlockId: lastBlock.id,
+			},
+		});
```
Comment (nazarhussain) [inline framework/src/modules/chain/loader.js]: Why this change appears in this PR?

File: framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+
+/*
+  DESCRIPTION: Creates meta table for BFT.
+
+  PARAMETERS: None
+*/
+
+CREATE TABLE IF NOT EXISTS "block_temp" (
+	"id" character varying(20) NOT NULL,
```
Comment (nazarhussain) [inline framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql]:
```suggestion
	"id" VARCHAR(20) PRIMARY KEY,
```

File: framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  "height" integer NOT NULL,
```
Comment (nazarhussain) [inline framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql]:
```suggestion
  "height" INT NOT NULL,
```

File: framework/src/modules/chain/transactions/transactions_handlers.js [hunk 12]
```diff
@@ -168,7 +168,7 @@ const applyTransactions = (storage, exceptions) => async (transactions, tx) => {
 				);
 			}
 			if (transactionResponse.status === TransactionStatus.OK) {
-				votes.apply(stateStore, transaction);
+				votes.apply(stateStore, transaction, exceptions);
```
Comment (nazarhussain) [inline framework/src/modules/chain/transactions/transactions_handlers.js]: Don't think this change belongs to scope of this PR.

**Comment by nazarhussain** [comment]
@SargeKhan Change the base branch on GitHub and then set base again, hopefully this will fix the problem on diff.

File: framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql (hunk 10 above)
Comment (michielmulders) [reply framework/src/modules/chain/migrations/sql/20190705101215_create_block_temp_table.sql]: Regarding Nazar's comments, the table name in the SQL files should be changes as well to `temp_block`

File: framework/src/modules/chain/components/storage/entities/block_temp.js (hunk 4 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/components/storage/entities/block_temp.js]: I think the `,` can be removed

File: framework/src/modules/chain/components/storage/entities/block_temp.js (hunk 6 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/components/storage/entities/block_temp.js]: Why is `_options` underscored?

File: framework/src/modules/chain/blocks/chain.js (hunk 1 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/blocks/chain.js]: I think `blockToJSON` or `convertBlockToJSON` suits as a better name, as `parse` is only meant to describe the act of reading and understanding the contents of the block

File: framework/src/modules/chain/components/storage/entities/block_temp.js (hunk 5 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/components/storage/entities/block_temp.js]: I think `json` should be `object` instead.

File: framework/src/modules/chain/blocks/chain.js [hunk 13]
```diff
@@ -195,11 +224,29 @@ const saveBlock = async (storage, block, tx) => {
 	parsedBlock.previousBlockId = parsedBlock.previousBlock;
 	delete parsedBlock.previousBlock;
 
-	parsedBlock.transactions.map(transaction => {
+	parsedBlock.transactions.forEach(transaction => {
```
Comment (limiaspasdaniel) [inline framework/src/modules/chain/blocks/chain.js]: Avoid using forEach if possible.

File: framework/src/modules/chain/blocks/chain.js [hunk 14]
```diff
@@ -195,11 +224,29 @@ const saveBlock = async (storage, block, tx) => {
...
+};
+
+/**
+ * Save block with transactions to database.
+ *
+ * @param {Object} block - Full normalized block
+ * @param {function} cb - Callback function
+ * @returns {Function|afterSave} cb - If SQL transaction was OK - returns safterSave execution, if not returns callback function from params (through setImmediate)
+ * @returns {string} cb.err - Error if occurred
+ */
```
Comment (limiaspasdaniel) [inline framework/src/modules/chain/blocks/chain.js]: These `jsdocs` are outdated. I don't know if it's the scope of this PR but maybe we could update them.

File: framework/src/modules/chain/blocks/chain.js (hunk 1 above)
Comment (SargeKhan) [reply framework/src/modules/chain/blocks/chain.js]: @nazarhussain I am keeping the same interface as `deleteLastBlock` since soon we might be replacing the functionality of deleteLastBlock with thiis function.

Comment (SargeKhan) [reply framework/src/modules/chain/blocks/chain.js]: @limiaspasdaniel what do you mean? I don't understand your point?

File: framework/src/modules/chain/blocks/chain.js (hunk 13 above)
Comment (SargeKhan) [reply framework/src/modules/chain/blocks/chain.js]: What do you suggest?

File: framework/src/modules/chain/blocks/chain.js (hunk 3 above)
Comment (SargeKhan) [reply framework/src/modules/chain/blocks/chain.js]: Why so?

File: framework/src/modules/chain/blocks/chain.js (hunk 1 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/blocks/chain.js]: Ignore my comment sorry, I read the name again and it indeed makes sense :)

File: framework/src/modules/chain/blocks/chain.js (hunk 13 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/blocks/chain.js]: `for (transaction of parsedBlock.transactions)`. However we haven't agreed about this yet so feel free to adopt it if the scope of the PR doesn't get too big.

File: framework/src/modules/chain/components/storage/entities/block_temp.js (hunk 6 above)
Comment (SargeKhan) [reply framework/src/modules/chain/components/storage/entities/block_temp.js]: simply using the existing interface defined in the rest of the application

File: framework/src/modules/chain/components/storage/sql/block_temp/create.sql (hunk 7 above)
Comment (SargeKhan) [reply framework/src/modules/chain/components/storage/sql/block_temp/create.sql]: Can you give me an example? I looked at create queries for other entities and they follow the same practice. I am not sure how you'd like me to change this particular query.

**Comment by limiaspasdaniel** [comment]
@SargeKhan something is wrong with the tests, could you please take a look?

File: framework/src/modules/chain/blocks/chain.js (hunk 3 above)
Comment (nazarhussain) [reply framework/src/modules/chain/blocks/chain.js]: Because I am not sure if `transaction.toJSON()` is preserving that custom attribute we assigned just above manually.

File: framework/src/modules/chain/blocks/chain.js (hunk 13 above)
Comment (nazarhussain) [reply framework/src/modules/chain/blocks/chain.js]: I posted earlier to Slack about this topic but it was not concluded. Let's conclude the topic outside PR and then all follow same pattern.

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+const sqlFiles = {
+	create: 'temp_block/create.sql',
+	delete: 'temp_block/delete.sql',
+	get: 'temp_block/get.sql',
+};
+
+/**
+ * Block temp
+ * @typedef {Object} temp_block
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]:
```suggestion
 * @typedef {Object} TempBlock
```

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+ * @property {string} id
+ * @property {number} height
+ * @property {object} fullBlock
+ */
+
+/**
+ * Round Filters
+ * @typedef {Object} filters.Round
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]:
```suggestion
 * @typedef {Object} filters.TempBlock
```

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+		this.sqlDirectory = path.join(path.dirname(__filename), '../sql');
+
+		this.SQLs = this.loadSQLFiles('temp_block', sqlFiles, this.sqlDirectory);
+	}
+
+	/**
+	 * Create temp_block row entry
+	 *
+	 * @param {Object} data
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]: * @param {TempBlock} data

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+		return this.adapter.executeFile(
+			this.SQLs.create,
+			{ createSet, fields },
+			{ expectedResultCount: 0 },
+			tx
+		);
+	}
+
+	async get(filters = {}, options = {}, tx = null) {
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]: Add JS doc for this function 

```
         * @param {filters. TempBlock|filters.TempBlock[]} [filters = {}]
	 * @param {Object} [options = {}] - Options to filter data
	 * @param {Number} [options.limit=10] - Number of records to fetch
	 * @param {Number} [options.offset=0] - Offset to start the records
	 * @param {string} [options.sort] - Sort key for transaction e.g. amount:asc, amount:desc
	 * @param {Boolean} [options.extended=false] - Get extended fields for entity
	 * @param {Object} tx - Database transaction object
	 * @return {Promise.< TempBlock[], Error>}
```

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+		return this._getResults(filters, options, tx);
+	}
+
+	async getOne(filters = {}, options = {}, tx = null) {
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]: Please here as well, add JS doc.

File: framework/src/modules/chain/components/storage/entities/temp_block.js [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+			params,
+			{ expectedResultCount },
+			tx
+		);
+	}
+
+	/**
+	 * Delete records with following conditions
+	 *
+	 * @param {filters.Round} filters
```
Comment (nazarhussain) [inline framework/src/modules/chain/components/storage/entities/temp_block.js]:
```suggestion
	 * @param {filters.TempBlock} filters
```
